Fix template parameter type leaking out

MatmaRex opened the pull request:
Fixes #4982.

One common operation with generic classes is, given a union type,
computing a map of template params to concrete types used in this
type, so that this map can be applied to another union type.

A problem occurs when the given variable's type is a generic type
that is missing some or all of the type parameters (e.g. determined
from a native PHP type annotation without a PHPDoc comment):
let's say an instance of `X` which is really `X<T>`. The computation
of the map would duly report that 'T' parameter is the `T` template
type, this would be applied elsewhere, but it was useless, as `T`
doesn't mean anything outside of the scope of the `X` class.

What should happen in this situation, in almost all cases, is that 'T'
should be substituted to indicate that we know nothing about the type.

I'm making this change in the first step of this operation (computing
the map, or `getTemplateParameterTypeMap`) rather than the second
(applying the map, `withTemplateParameterTypeMap`), because it turns
out that the second one should, in almost all cases, not change types
that aren't listed in the map; for example, we often want to map a
class's template params in an inherited method, but leave the method's
own template params unchanged.

MatmaRex committed:
Fix template parameter type leaking out

Fixes #4982.

One common operation with generic classes is, given a union type,
computing a map of template params to concrete types used in this
type, so that this map can be applied to another union type.

A problem occurs when the given variable's type is a generic type
that is missing some or all of the type parameters (e.g. determined
from a native PHP type annotation without a PHPDoc comment):
let's say an instance of `X` which is really `X<T>`. The computation
of the map would duly report that 'T' parameter is the `T` template
type, this would be applied elsewhere, but it was useless, as `T`
doesn't mean anything outside of the scope of the `X` class.

What should happen in this situation, in almost all cases, is that 'T'
should be substituted to indicate that we know nothing about the type.

I'm making this change in the first step of this operation (computing
the map, or `getTemplateParameterTypeMap`) rather than the second
(applying the map, `withTemplateParameterTypeMap`), because it turns
out that the second one should, in almost all cases, not change types
that aren't listed in the map; for example, we often want to map a
class's template params in an inherited method, but leave the method's
own template params unchanged.

MatmaRex commented on tests/files/expected/0962_iteratoraggregate_template.php.expected:
```diff
@@ -1,6 +1,6 @@
 %s:49 PhanDebugAnnotation @phan-debug-var requested for variable $iterator - it has union type \Iterator<int,\UserModel>
 %s:49 PhanDebugAnnotation @phan-debug-var requested for variable $users - it has union type \Collection<\UserModel>(real=\Collection)
-%s:51 PhanDebugAnnotation @phan-debug-var requested for variable $key - it has union type int
+%s:51 PhanDebugAnnotation @phan-debug-var requested for variable $key - it has union type int(real=int)
```
IDK where this real type comes from.

Daimona replied:
Traced it to https://github.com/phan/phan/blob/41d5b918bc2075d280847136cb31608cd482bc8a/src/Phan/Language/Type.php#L2478-L2483

The `withTemplateParameterTypeMap` previously returned `\Iterator<int,T>`, whereas with your patch, it now returns `\Iterator<int,(empty)>`. Then, `withoutTemplateTypeRecursive` drops any type containing templates, so previously it would empty the union type, whereas with your change it survives and gets used later. And then there's an unrelated bug in UnionType, which makes it use non-real types as real; I fixed that in #4985.

This, however, also makes me wonder if #4982 could be caused by a missing `withoutTemplateTypeRecursive` call somewhere. The whole template type handling confuses me. Maybe I misunderstood what `withTemplateParameterTypeMap` does, and it is after all allowed to leave unresolved template types in the return value, if they aren't contained in the map. And then the caller is responsible for removing them.

MatmaRex replied:
I don't like `withoutTemplateTypeRecursive`. It's only used in 2 methods dealing with iterable types, added in 7d47aeb64eab920222612d43f6bb85e7563b50ce and a5a988808ff3b04d921452de8b9338433e20e27b, and it feels like a workaround for something (maybe for the same bug as I'm trying to fix).

Maybe I'll see if it's possible to remove these uses…

Daimona replied:
Yeah, it is very confusing. I'm not sure what the best fix would be...

Daimona commented:
I also found this: https://github.com/phan/phan/blob/41d5b918bc2075d280847136cb31608cd482bc8a/src/Phan/Language/Type/TemplateType.php#L129-L136

The documentation says all templates will be resolved, but then it uses `$this->asPHPDocUnionType()`, which effectively returns the template (unresolved) type. That seems incorrect, so maybe the `UnionType::empty()` should be used there instead? I don't know if that's going to have any differences in practice though.

MatmaRex commented:
> I also found this: `TemplateType::withTemplateParameterTypeMap()`
> The documentation says all templates will be resolved, but then it uses `$this->asPHPDocUnionType()`, which effectively returns the template (unresolved) type. That seems incorrect, so maybe the `UnionType::empty()` should be used there instead? I don't know if that's going to have any differences in practice though.

Yep, that looks like a better place!

In practice this is almost the same, as `withTemplateParameterTypeMap()` and `getTemplateParameterTypeMap()` are almost always used together. The only exception I found is `UnionType::usesTemplateType()`, which relies on the current behavior of `withTemplateParameterTypeMap()`. There should be a better way to write that function.

MatmaRex commented:
After banging my head against this for a few hours (making that change generates all kinds of interesting test failures), I think what I really want is for `$this` to have the type `static<...>` (with all the class template parameters) instead of just `static`. That should fix my particular case, but I wonder what else it might break. I'll give it a try.

MatmaRex commented:
To expand on the idea from the previous comment, after banging my head against it for a few more hours, and with better understanding of what I want to do:

The initial idea here is right. When we're substituting template types with concrete types, and any of them were not provided, we should substitute them with _something_. Substituting them with the empty union type (which I understand to be Phan's version of `any`) seems ideal, as it will avoid emitting bogus issues elsewhere. Instead, we should probably emit just one issue at the definition site, reporting that a type parameter was missing.

However, I found 2 places where Phan depends on the current behavior. One of them was fixed in #4988. The other is in type checking the `static` and `self` types: if you have a generic class like `Pair<A, B>`, and one of its methods should return `B` but actually returns  `$this->a`, this would previously emit an issue about `A` not matching `B`, but now it won't because `$this` is of type `static`, and `$this->a` is now `any` instead of `A`. That's why I think that `$this` needs to be of type `static<A, B>`, so that `$this->a` may be `A`, and this issue will once again be emitted. (And likewise for anything that is currently an unparameterized `static` or `self`, like `new static(...)`.)

I'm about 10% of the way towards implementing that. Let me know if it seems like a bad idea.

This may lead to trouble when a generic class extends another generic class and substitutes some template parameters with other template parameters. I haven't gotten far enough to know whether type checks will work when they involve e.g. a `MyPair<X, Y>` with `@extends Pair<A, B>`, with `$this->a` of type `A` inherited from a parent and something of type `X` or `Y` from the current class.

Daimona commented:
> I'm about 10% of the way towards implementing that. Let me know if it seems like a bad idea.

I'm sure that by now, you know much, much more than me about generics ;)

The only thing I could add is that, if that change turns out to be too difficult, maybe there'd be a simpler fix using `withoutTemplateTypeRecursive`? Although I guess it would need to be changed so that it turns `Foo<T>` into just `Foo`, instead of dropping the type entirely. And I don't know if that would really end up being easier than your proposal.

MatmaRex commented:
I've spent some time trying to better understand how Phan handles generic types in order to eventually figure out whether this is a good idea or not. (Hence my other pull requests. For some of the parts I couldn't understand, it turned out that they in fact made no sense. :) )



> The initial idea here is right. When we're substituting template types with concrete types, and any of them were not provided, we should substitute them with _something_.

I think this is mostly still true, but there may be some cases where we want to leave the template types unsubst'ed, e.g. maybe when a class with templates has a method that also has templates, and we need to subst them one set at a time. Or maybe we can merge the sets. I need to look into that more.


> Substituting them with the empty union type (which I understand to be Phan's version of `any`) seems ideal, as it will avoid emitting bogus issues elsewhere.

Apparently both the empty union type and `mixed` are Phan's version of `any` (by which I mean "unknown type, permit all casts and operations, do not emit any warnings about it"). See also #3984 about that fact that `mixed` never causes any warnings, and whether it should.



> Instead, we should probably emit just one issue at the definition site, reporting that a type parameter was missing.

I tried this and got lots of bogus issues. It's not easy to tell where the definition site is when all you have is a `Type`. I think I'll drop this idea.


> However, I found 2 places where Phan depends on the current behavior. One of them was fixed in #4988. The other is…

Method templates on class templates (mentioned above) may be another one, not sure. There are also things that break in `asExpandedTypes()`, but they may just be hidden bugs masked by other bugs, I don't understand that family of functions well enough just yet.


> That's why I think that `$this` needs to be of type `static<A, B>`, so that `$this->a` may be `A`, and this issue will once again be emitted. (And likewise for anything that is currently an unparameterized `static` or `self`, like `new static(...)`.)

This is actually wrong, because we don't even know how many type parameters `static` has (subclasses can add or remove them). Maybe they can be added later when resolving `static` to a class type, or maybe this is a bad idea.


> This may lead to trouble when a generic class extends another generic class and substitutes some template parameters with other template parameters. I haven't gotten far enough to know whether type checks will work when they involve e.g. a `MyPair<X, Y>` with `@extends Pair<A, B>`, with `$this->a` of type `A` inherited from a parent and something of type `X` or `Y` from the current class.

I added tests in #4998 and this is definitely broken already and isn't made worse by these changes.

MatmaRex commented:
> The only thing I could add is that, if that change turns out to be too difficult, maybe there'd be a simpler fix using `withoutTemplateTypeRecursive`? Although I guess it would need to be changed so that it turns `Foo<T>` into just `Foo`, instead of dropping the type entirely. And I don't know if that would really end up being easier than your proposal.

I tried to figure out the purpose of `withoutTemplateTypeRecursive` and haven't found any, it just breaks things. I removed it in the one place it was used. :)

----

My idea right now, in order to get this done within days and not weeks, is to split `withTemplateParameterTypeMap()` into a few functions (or maybe add a param) that behave differently when a mapping for a template type is not provided: one will substitute empty union types, another will keep the template type, and maybe another will throw an error (if I find any places where I can do that with enough confidence). It feels a bit sloppy, but there seem to be reasonable use cases for both behaviors, and there seem to be places in the code that depend on the current behavior for reasons that are hard to understand.

MatmaRex committed:
Fix support for nested generics in template types

MatmaRex committed:
Map inherited template types in some more edge cases

Sometimes, the template types of properties and method parameters
that we've so carefully mapped would be overwritten later.

Their types are based on their PHPDoc and default value types, and
since the default values can depend on constants from classes that may
not be loaded yet when the code is parsed, the processing of *both*
PHPDoc and default values is deferred. (Property::getFutureUnionType(),
FunctionTrait::addParamsToScopeOfFunctionOrMethod())

In order to support the case where it happens after the class
inheritance, map inherited template types in the internal
representation of PHPDoc comments too.

This is ugly, I wish there was a better way to avoid this problem.

MatmaRex committed:
Warn about unresolved template types if they both come from the same class scope

MatmaRex committed:
Check generic types during property assignments

MatmaRex committed:
Keep Method::hasTemplateType() more in sync with the params

There were some cases where we would replace all template types with
concrete types, but the flag on the method was not updated. This was
mostly harmless, but possibly inefficient.

There might have also been cases where a method was created with
template types in a more complex manner (e.g. nested deeper inside a
type parameter, or only in the unresolved PHPDoc type), but we don't
have a test case for that.

MatmaRex committed:
Method: Simplify resolveTemplateType()

The previous code tried to resolve template types using methods
found on parent classes and interfaces of the class we're examining.
This doesn't work reliably, because they can take completely different
template parameters (and we don't support generic interfaces anyway).

It shouldn't be needed in the first place: much like in 0640175de8,
the actual resolution of inherited template parameter names happens
in Clazz when copying the inherited method.

MatmaRex committed:
Clazz: Move Method/Property cloning code to their classes

Method already had a similar method, used when cloning method for a
one-time use of a generic method with known types.

The code for Property had a try...catch around it, but IssueException
can never be thrown there.

Somehow this causes Phan to detect a new issue with documentation in
its own code. No idea how that happened.

MatmaRex committed:
ParameterTypesAnalyzer: Use Method::cloneWithTemplateParameterTypeMap()

MatmaRex committed:
Handle extracting template types out of variadic parameters

When no variadic arguments were provided, the value representing the
variadic parameter is of type `never`. This could be used to improve
analysis later.

MatmaRex commented:
Now depends on #5013, wherein I've made many other improvements to the handling of generics. I am now pretty sure that this is correct, and I found exactly 1 case where it wasn't and the previous behavior needs to be kept.
